### DISCUSSION TITLE
往node.js的hono引入Reader Monad

### DISCUSSION THREAD

**Discussion by kalxd** [just-talks/tech-talks#72] [3.分享创造]
[Hono](https://hono.dev/)有点类似express.js，只有路由能力，以及中间件功能。
这种甚至不能称得上框架，只能是一个库，很多功能需要从头造起，其中一个就是依赖注入。例如我们追踪整条请求链路，一般会为每个请求创建唯一id，这个id伴随整个请求周期，而且会经过很多业务代码、函数；如果都要显示传id，十分枯燥、烦琐。

现有依赖注入方式大都通过装饰器注入，装饰器仅在运行时检查，依赖能否注入，需要运行起来才知道；Reader Monad可以认为是编译期的依赖注入，类型安全的依赖注入。

reader可以等价于一元函数：`(r: T) => R`。所谓依赖注入，也就是在`r`这个巨大环境变量下操作着业务。

这里我用typescript实现了一个简易版的reader monad：

```typescript
const getHandler = reader<Context<Env, any, BlankInput>, string | undefined>(helper => {
	const ctx = helper.ask(); // 获取当前上下文，即`r`
	return ctx.req.header("Content-Type");
});

const testHandler = reader<Context<Env, "/test", BlankInput>, string>(helper => {
	const header = helper.bindFrom(getHandler); // 引用其他依赖结果
	if (!header) {
		return "not founed";
	}

	return header;
});

app.post("/test", async ctx => {
	const r = runReader(testHandler, ctx);
	return ctx.json(r);
});
```

上文的`testHandler`使用`helper.bindFrom(getHandler)`，这里不用显式传参给`getHandler`就能获取`getHandler`的结果。也就是我们通过`bindFrom`就能实现依赖调用问题。

以上的代码都不需要显示传参，一切都包含在`helper`里，它自动处理了这一切。

实现代码：https://github.com/kalxd/lickitung/blob/master/src/lib/reader.ts

**Comment by kalxd** [comment]
不是，为啥贴了前端标签？
我是为了后端的依赖注入才搞了这么个玩意儿。

**Reply by adoyle-h** [reply]
呃，那换成 Web 标签

**Reply by kalxd** [reply]
痛心，十分痛心！这才短短几年，nodejs已经成为前端代名词了！

**Reply by adoyle-h** [reply]
不是一直都是吗 （笑）
